### ISSUE TITLE
leadingDistribution not work in TextField

### ISSUE THREAD

**Issue by xuzhongpeng** [flutter/flutter#115721]
## Steps to Reproduce

`leadingDistribution: TextLeadingDistribution.even` work fine in Text widget but TextField.
<img width="380" alt="image" src="https://user-images.githubusercontent.com/24218764/202902231-b90b3d3b-bef0-4eb1-a82d-49d58f75ee39.png">
<details>
<summary>Code sample</summary>

```dart
Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children:const <Widget>[
            Text(
              'You have pushed the button',
              style: TextStyle(
                  height: 10,
                  fontSize: 20,
                  leadingDistribution: TextLeadingDistribution.even),
            ),
            TextField(
              style: TextStyle(
                height: 10,
                fontSize: 20,
                leadingDistribution: TextLeadingDistribution.even,
              ),
            ),
          ],
        ),
      )
```

</details>

<details>
  <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel flutter-2.13-candidate.0, 3.0.5, on macOS 12.4 21F79 darwin-x64, locale zh-Hans-CN)
    • Flutter version 3.0.5 at /Users/admin/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (4 months ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /usr/local/Caskroom/android-sdk/4333796
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /usr/local/Caskroom/android-sdk/4333796/
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS (Xcode 13.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! CocoaPods 1.10.0 out of date (1.11.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the
        Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio
    • Android Studio at /Applications/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[!] Android Studio
    • Android Studio at /Users/admin/Downloads/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] VS Code (version 1.69.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.48.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 12.4 21F79 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 3 categories.
```
</details>

Related to #72521 #23202 #96713

**Comment by danagbemava-nc** [comment]
Reproducible using the code sample provided. 

<img width="350" alt="Screenshot 2022-11-21 at 08 45 08" src="https://user-images.githubusercontent.com/88313112/203006398-7ad43c33-0bd8-44f5-9f87-c8f9315eeb1a.png">


<details>
<summary>complete sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  final String title;

  const MyHomePage({super.key, required this.title});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: const <Widget>[
          Text(
            'You have pushed the button',
            style: TextStyle(height: 10, fontSize: 20, leadingDistribution: TextLeadingDistribution.even),
          ),
          TextField(
            style: TextStyle(
              height: 10,
              fontSize: 20,
              leadingDistribution: TextLeadingDistribution.even,
            ),
          ),
        ],
      ),
    );
  }
}
```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.3.8, on macOS 13.0.1 22A400 darwin-arm, locale en-GB)
    • Flutter version 3.3.8 on channel stable at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 52b3dc25f6 (12 days ago), 2022-11-09 12:09:26 +0800
    • Engine revision 857bd6b74c
    • Dart version 2.18.4
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio
    • Android Studio at /Applications/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[!] Android Studio
    • Android Studio at /Applications/Android Studio Preview 2.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] Android Studio (version 2021.3)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] Android Studio (version 2021.3)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9014738/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.73.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.52.0

[✓] Connected device (4 available)
    • TA 1021 (mobile)       • PLEGAR1790423460                     • android-arm64  • Android 9 (API 28)
    • iPhone 14 Pro (mobile) • 4F72110C-F38B-4CF9-93C4-4D6042148D28 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-1 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 13.0.1 22A400 darwin-arm
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 2 categories.
```

```
[!] Flutter (Channel master, 3.6.0-6.0.pre.29, on macOS 13.0.1 22A400 darwin-arm64, locale en-GB)
    • Flutter version 3.6.0-6.0.pre.29 on channel master at /Users/nexus/dev/sdks/flutters
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ee9bc784b5 (8 hours ago), 2022-11-20 19:32:24 -0500
    • Engine revision 271461837e
    • Dart version 2.19.0 (build 2.19.0-406.0.dev)
    • DevTools version 2.19.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio
    • Android Studio at /Applications/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[!] Android Studio
    • Android Studio at /Applications/Android Studio Preview 2.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] Android Studio (version 2021.3)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] Android Studio (version 2021.3)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9014738/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.73.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.52.0

[✓] Connected device (4 available)
    • TA 1021 (mobile)       • PLEGAR1790423460                     • android-arm64  • Android 9 (API 28)
    • iPhone 14 Pro (mobile) • 4F72110C-F38B-4CF9-93C4-4D6042148D28 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-1 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 13.0.1 22A400 darwin-arm64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 3 categories.

```
</details>

**Comment by Malarg** [comment]
I'm trying to fix it

**Comment by justinmc** [comment]
@Malarg Are you still looking at fixing this?

CC @LongCatIsLooong

**Comment by Malarg** [comment]
Yes, I am. I already found the problem. I'll create pull request in a few days

**Comment by justinmc** [comment]
Awesome!  Feel free to request review from me and @LongCatIsLooong.

**Comment by Malarg** [comment]
@justinmc  At the end of my investigation I believe, that it's an engine issue. 
The core problem in the `strutStyle` property. It responsible for layout text by its height. In the given sample code it doesn't set in `TextField` and its generated in `TextEditingController's` `strutStyle` getter:
```
  StrutStyle get strutStyle {
    if (_strutStyle == null) {
      return StrutStyle.fromTextStyle(style, forceStrutHeight: true);
    }
    return _strutStyle!.inheritFromTextStyle(style);
  }
```
It's the key difference between `TextField` and `Text` mechanisms of working with `strutStyle` - in `Text` in won't be generated if it has not be set when `Text` created.
For checking theory that fix should be made at engine, I changed sample code and added `strutStyle` to `Text` and it became looks buggy too. We can't affect on painting text based on `StrutStyle`, because flutter just build `Paragraph` and passes it next to engine. 

I am sorry in advance, if I am wrong: it my first experience contributing to flutter. Please point to inaccuracies in my thoughts and give me a chance to fix it if I am not right and problem is in framework. 

Changed sample with result screenshot
<img width="273" alt="Снимок экрана 2022-12-04 в 14 48 57" src="https://user-images.githubusercontent.com/17158960/205486834-36e79a42-2795-4339-815a-bc81784a11f5.png">

```
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  final String title;

  const MyHomePage({super.key, required this.title});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    const textStyle = TextStyle(
      height: 10,
      fontSize: 20,
      leadingDistribution: TextLeadingDistribution.even,
      backgroundColor: Colors.blue,
    );
    return Scaffold(
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          const SizedBox(
            height: 150,
          ),
          _RedContainer(
            child: Text(
              'Some text',
              style: textStyle,
              strutStyle: StrutStyle.fromTextStyle(textStyle),
            ),
          ),
          const SizedBox(height: 4),
          _RedContainer(
            child: TextField(
              controller: TextEditingController(text: 'Some text'),
              style: textStyle,
            ),
          ),
        ],
      ),
    );
  }
}

class _RedContainer extends StatelessWidget {
  final Widget child;
  const _RedContainer({required this.child});

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.all(4),
      color: Colors.red,
      child: child,
    );
  }
}
```

P.S. The second thought I have - `Text` and `TextField` should work the same way with `StrutStyle`. Can I create an issue fixing this?

**Comment by jason-simmons** [comment]
This is happening in SkParagraph due to discrepancies between the run's `correctAscent/Descent` (which is adjusted based on `TextLeadingDistribution`) and the strut's ascent/descent (which is not adjusted).

SkParagraph's `TextLeadingDistribution.even` implementation works as expected in a simple case with no strut:
* `Run::calculateMetrics` will calculate the `correctAscent/Descent` so that the text is vertically centered within the run's height
* `InternalLineMetrics::add` will accumulate the `correctAscent/Descent` into the `InternalLineMetrics`
* `TextLine` painting code calls `InternalLineMetrics::runTop` to find the position where the text should be drawn.  `runTop` calculates this based on `InternalLineMetrics.fAscent` and the run's `correctAscent`

But if a strut is present, then the ascent of the `InternalLineMetrics` will diverge from the run's `correctAscent`.

Layout code will call `InternalLineMetrics::updateLineMetrics` to apply the strut.  `updateLineMetrics` will replace fields like `fAscent` with the strut's values.  Later, `runTop` will combine the `InternalLineMetrics.fAscent` (from the strut) with the `run.correctAscent()` (which was calculated before the strut was applied) and will return an incorrect position.

@Rusino Not sure what would be the simplest way to handle this.

One option would be representing `TextLeadingDistribution.even` as another offset that is separate from the height-multiplied ascent/descent so that adjusted and unadjusted ascent/descent values do not get mixed in the `InternalLineMetrics`.

**Comment by mateusfccp** [comment]
Considering that this problem seems to be related to Skia (https://github.com/flutter/flutter/issues/115721#issuecomment-1355944179), is this fixed if we use Impeller?

**Comment by LongCatIsLooong** [comment]
No longer reproducible.

**Comment by mateusfccp** [comment]
> No longer reproducible.

Do you mean it's not reproducible in Impeller anymore? Or in general?

I can easily reproduce using Skia.

**Comment by LongCatIsLooong** [comment]
I believe this is a duplicate of https://github.com/flutter/flutter/issues/142969. 

@mateusfccp I don't think this has to do with Impeller as it doesn't do text layout. Can you post a repro / create a new issue if you're still experiencing this issue?

**Comment by mateusfccp** [comment]
> I believe this is a duplicate of https://github.com/flutter/flutter/issues/142969. 
> 
> @mateusfccp I don't think this has to do with Impeller as it doesn't do text layout. Can you post a repro / create a new issue if you're still experiencing this issue?

Sorry, we are using 3.19. If it was fixed in 3.22 we still have to test.

When we do it, if I can still reproduce the issue, I will create a new issue.
